components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -257,10 +262,29 @@ abstract class AbstractFetchDownloadService : Service() {
         context.registerReceiver(broadcastReceiver, filter)
     }
 
-    private fun displayOngoingDownloadNotification(download: DownloadState) {
+    /***
+     * Android rate limits notifications being sent, so we must send them on a delay so that
+     * notifications are not dropped
+     */
+    private fun beginOngoingNotificationProgressUpdates(downloadID: Long) {
+        notificationTimer = Timer()
+        val downloadJobState = downloadJobs[downloadID] ?: return
+        val download = downloadJobState.state
+
+        val timerTask = object : TimerTask() {
+            override fun run() {
+                displayOngoingDownloadNotification(download, downloadJobState.currentBytesCopied)
+            }
+        }
+
+        notificationTimer.scheduleAtFixedRate(timerTask, 0, PROGRESS_UPDATE_INTERVAL)
```

Comment:
I am a bit confused about the lifetime of the tasks. We schedule one task for every download here. But in `startDownoadJob()` we cancel all tasks on the timer. So I am wondering:
* Doesn't the `cancel()` call in `startDownoadJob()` cause all other notifications to get stale except the new one?
* We never cancel a specific task after the download is done right? Only if the service stops and the `Timer` object gets garbage collected then they get away (but the docs are not very specific on when that happens).

I wonder if we could use a single Kotlin Coroutine to update all notifications, something like:

```Kotlin
val notificationUpdateScope = MainScope()

override fun onCreate() {
  // [..]
  notificationUpdateScope.launch {
     delay(...)
     updateProgress()
  }
}

override fun onDestroy() {
   // [..]
   notificationUpdateScope.cancel()
}

fun updateProgress() {
   // Update progress of all notifications.
}
```

Does this make sense? I think with a single coroutine it is much easier to understand its lifetime and avoiding multiple zombie threads being left behind. What do you think?

Reply:
All great points. I have updated this PR with the ideas you suggested.